src/hotspot/share/opto/superword.cpp

```diff
@@ -2572,8 +2572,8 @@ static bool can_subword_truncate(Node* in, const Type* type) {
     return false;
   }
 
-  // Vector nodes should not truncate.
-  if (type->isa_vect() != nullptr || type->isa_vectmask() != nullptr || in->is_Reduction()) {
+  // Vector nodes and casts should not truncate.
+  if (type->isa_vect() != nullptr || type->isa_vectmask() != nullptr || in->is_Reduction() || in->is_ConstraintCast()) {
```

Comment:
Why should we not truncate a CastII? What can go wrong?

Reply:
My thinking was that since casts specifically change the type of a node, they may not be safe to truncate. In practice it might not matter because after the CastII pack is created, it's discarded because there is no backend implementation for vectorized CastII. I've opted to mark them as non-truncating to stay on the safer side.

Reply:
I see. Ok. Can you add a comment to the code for that?
Because imagine we come along later and actually implement a backend vectorized version of CastII (no-op?). Maybe because we implement if-conversion. Then it would be nice to know if this was just a "to be on the safe side" check, or if it would run into issues when removed.

Reply:
Because the current comment says "should not truncate". That sounds more strong than "to be on the safe side".

Reply:
I think this is fair, I've pushed a commit that changes the comment wording. Let me know what you think!

Reply:
We have `CastVV` which should be the packed version of `CastII`. The thing that is I think the most difficult is to properly wire it. In the simplest situation of all elements in the pack having the same control input can be easily handled, though.